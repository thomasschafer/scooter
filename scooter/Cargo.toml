[package]
name = "scooter"
version = "0.5.0"
edition = "2021"
authors = ["thomasschafer97@gmail.com"]
license = "MIT"
description = "Interactive find and replace in the terminal"
readme = "../README.md"
homepage = "https://github.com/thomasschafer/scooter"
repository = "https://github.com/thomasschafer/scooter"
keywords = ["cli", "find", "search", "replace"]
categories = ["command-line-utilities"]

[dependencies]
anyhow = "1.0.97"
clap = { version = "4.5.32", features = ["derive"] }
content_inspector = "0.2.4"
crossterm = { version = "0.27", features = ["event-stream"] }
etcetera = "0.8.0"
fancy-regex = "0.14.0"
futures = "0.3.31"
ignore = "0.4.23"
itertools = "0.14.0"
log = "0.4.26"
lru = "0.14.0"
parking_lot = "0.12.3"
ratatui = "0.27.0"
regex = "1.11.1"
serde = "1.0.219"
similar = "2.7.0"
simple-log = "2.1.2"
syntect = "5.2.0"
tempfile = "3.18.0"
tokio = { version = "1.44.2", features = ["full"] }
toml = "0.8.20"

[dev-dependencies]
rand = "0.8.5"
serial_test = "3.2.0"

[lib]
name = "scooter"
path = "src/lib.rs"

[lints.rust]
rust_2018_idioms = "warn"
