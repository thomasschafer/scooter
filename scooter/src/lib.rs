pub mod app_runner;
pub mod headless;
pub mod logging;
pub mod tui;
pub mod ui;
