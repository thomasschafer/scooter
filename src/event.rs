// use crossterm::event::{self, Event as CrosstermEvent, KeyEvent, MouseEvent};
// use futures::StreamExt;
// use std::path::PathBuf;
// use tokio::sync::mpsc;

// use crate::app::ReplaceState;

// #[derive(Clone, Debug, PartialEq, Eq)]
// pub enum ReplaceResult {
//     Success,
//     Error(String),
// }

// #[derive(Clone, Debug, PartialEq, Eq)]
// pub struct SearchResult {
//     pub path: PathBuf,
//     pub line_number: usize,
//     pub line: String,
//     pub replacement: String,
//     pub included: bool,
//     pub replace_result: Option<ReplaceResult>,
// }

// #[derive(Debug)]
// pub enum AppEvent {
//     Rerender,
//     PerformSearch,
// }

// #[derive(Debug)]
// pub enum BackgroundProcessingEvent {
//     AddSearchResult(SearchResult),
//     SearchCompleted,
//     ReplacementCompleted(ReplaceState),
// }

// #[derive(Debug)]
// pub enum Event {
//     Key(KeyEvent),
//     App(AppEvent),
//     #[allow(dead_code)]
//     Mouse(MouseEvent),
//     #[allow(dead_code)]
//     Resize(u16, u16),
// }

// #[derive(Debug)]
// pub struct EventHandler {
//     pub receiver: mpsc::UnboundedReceiver<Event>,
//     pub app_event_sender: mpsc::UnboundedSender<AppEvent>,
// }

// #[derive(Debug)]
// pub struct EventHandlingResult {
//     pub exit: bool,
//     pub rerender: bool,
// }

// impl EventHandler {
//     pub fn new() -> Self {
//         let (sender, receiver) = mpsc::unbounded_channel();
//         let (app_event_sender, mut app_event_receiver) = mpsc::unbounded_channel();
//         tokio::spawn(async move {
//             let mut reader = event::EventStream::new();
//             loop {
//                 tokio::select! {
//                     Some(Ok(evt)) = reader.next() => {
//                         match evt {
//                             CrosstermEvent::Key(key) => {
//                                 if key.kind == crossterm::event::KeyEventKind::Press {
//                                     sender.send(Event::Key(key)).unwrap();
//                                 }
//                             },
//                             CrosstermEvent::Mouse(mouse) => {
//                                 sender.send(Event::Mouse(mouse)).unwrap();
//                             },
//                             CrosstermEvent::Resize(x, y) => {
//                                 sender.send(Event::Resize(x, y)).unwrap();
//                             },
//                             _ => {}
//                         }
//                     }
//                     Some(app_evt) = app_event_receiver.recv() => {
//                         sender.send(Event::App(app_evt)).unwrap();
//                     }
//                     else => break,
//                 };
//             }
//         });
//         Self {
//             receiver,
//             app_event_sender,
//         }
//     }
// }

// impl Default for EventHandler {
//     fn default() -> Self {
//         Self::new()
//     }
// }
