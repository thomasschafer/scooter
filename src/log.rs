use std::path::{Path, PathBuf};

use etcetera::base_strategy::{choose_base_strategy, BaseStrategy};

pub fn _cache_dir() -> PathBuf {
    let strategy = choose_base_strategy().expect("Error when finding cache directory");
    let mut path = strategy.cache_dir();
    path.push("scout");
    path
}

pub fn _default_log_file() -> PathBuf {
    _cache_dir().join("scout.log")
}

fn _make_parent_dir(path: &Path) {
    if let Some(parent) = path.parent() {
        if !parent.exists() {
            std::fs::create_dir_all(parent).ok();
        }
    }
}
