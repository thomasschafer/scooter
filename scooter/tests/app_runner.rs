use anyhow::bail;
use crossterm::event::{Event as CrosstermEvent, KeyCode, KeyEvent, KeyModifiers};
use futures::Stream;
use log::LevelFilter;
use ratatui::backend::TestBackend;
use scooter::{
    app_runner::{AppConfig, AppRunner},
    test_with_both_regex_modes, test_with_both_regex_modes_and_fixed_strings,
};
use std::{io, path::Path, pin::Pin, task::Poll};
use tokio::{
    sync::mpsc::{self, UnboundedReceiver, UnboundedSender},
    task::JoinHandle,
    time::{sleep, Duration, Instant},
};

mod utils;

struct TestEventStream(UnboundedReceiver<CrosstermEvent>);

impl TestEventStream {
    fn new() -> (UnboundedSender<CrosstermEvent>, Self) {
        let (sender, receiver) = mpsc::unbounded_channel();
        (sender, Self(receiver))
    }
}

impl Stream for TestEventStream {
    type Item = Result<CrosstermEvent, io::Error>;

    fn poll_next(
        self: Pin<&mut Self>,
        cx: &mut std::task::Context<'_>,
    ) -> Poll<Option<Self::Item>> {
        self.get_mut()
            .0
            .poll_recv(cx)
            .map(|opt| opt.map(Result::Ok))
    }
}

async fn wait_for_text(
    snapshot_rx: &mut UnboundedReceiver<String>,
    text: &str,
    timeout_ms: u64,
) -> anyhow::Result<String> {
    let timeout = Duration::from_millis(timeout_ms);
    let start = Instant::now();
    let mut last_snapshot = None;

    while start.elapsed() <= timeout {
        tokio::select! {
            snapshot = snapshot_rx.recv() => {
                match snapshot {
                    Some(s) if s.contains(text) => return Ok(s),
                    Some(s) => { last_snapshot = Some(s); },
                    None => bail!("Channel closed while waiting for text: {text}"),
                }
            }
            _ = sleep(timeout - start.elapsed()) => {
                break;
            }
        }
    }

    let formatted_snapshot = match last_snapshot {
        Some(snapshot) => &format!("Current buffer snapshot:\n{snapshot}"),
        None => "No buffer snapshots recieved",
    };
    bail!("Timeout waiting for text: {text}\n{formatted_snapshot}")
}

type TestRunner = (
    JoinHandle<()>,
    UnboundedSender<CrosstermEvent>,
    UnboundedReceiver<String>,
);

fn build_test_runner(directory: Option<&Path>, advanced_regex: bool) -> anyhow::Result<TestRunner> {
    let backend = TestBackend::new(80, 24);
    let config = AppConfig {
        directory: directory.map(|d| d.to_str().unwrap().to_owned()),
        hidden: false,
        advanced_regex,
        log_level: LevelFilter::Warn,
    };

    let (event_sender, event_stream) = TestEventStream::new();
    let (snapshot_tx, snapshot_rx) = mpsc::unbounded_channel();

    let mut runner =
        AppRunner::new(config, backend, event_stream)?.with_snapshot_channel(snapshot_tx);
    runner.init()?;

    let run_handle = tokio::spawn(async move {
        runner.run_event_loop().await.unwrap();
    });

    Ok((run_handle, event_sender, snapshot_rx))
}

async fn shutdown(
    event_sender: UnboundedSender<CrosstermEvent>,
    run_handle: JoinHandle<()>,
) -> anyhow::Result<()> {
    event_sender.send(CrosstermEvent::Key(KeyEvent::new(
        KeyCode::Esc,
        KeyModifiers::empty(),
    )))?;

    run_handle.await?;

    Ok(())
}

fn send_key(key: KeyCode, event_sender: &UnboundedSender<CrosstermEvent>) {
    event_sender
        .send(CrosstermEvent::Key(KeyEvent::new(
            key,
            KeyModifiers::empty(),
        )))
        .unwrap();
}

fn send_chars(word: &str, event_sender: &UnboundedSender<CrosstermEvent>) {
    word.chars()
        .for_each(|key| send_key(KeyCode::Char(key), event_sender));
}

#[tokio::test]
async fn test_search_current_dir() -> anyhow::Result<()> {
    let (run_handle, event_sender, mut snapshot_rx) = build_test_runner(None, false)?;

    wait_for_text(&mut snapshot_rx, "Search text", 10).await?;

    send_key(KeyCode::Enter, &event_sender);

    wait_for_text(&mut snapshot_rx, "Still searching", 500).await?;

    wait_for_text(&mut snapshot_rx, "Search complete", 1000).await?;

    shutdown(event_sender, run_handle).await
}

test_with_both_regex_modes!(
    test_search_and_replace_simple_dir,
    |advanced_regex| async move {
        let temp_dir = &create_test_files!(
            "dir1/file1.txt" => {
                "This is some test content before 123",
                "  with some spaces at the start",
                "and special ? characters 1! @@ # and number 890",
                "        some    tabs  and - more % special   **** characters ())",
            },
            "file2.py" => {
                "from datetime import datetime as dt, timedelta as td",
                "def mix_types(x=100, y=\"test\"): return f\"{x}_{y}\" if isinstance(x, int) else None",
                "class TestClass:",
                "    super_long_name_really_before_long_name_very_long_name = 123",
                "    return super_long_name_really_before_long_name_very_long_name",
                "test_dict = {\"key1\": [1,2,3], 123: \"num key\", (\"a\",\"b\"): True, \"before\": 1, \"test-key\": None}",
            },
        );

        let (run_handle, event_sender, mut snapshot_rx) =
            build_test_runner(Some(temp_dir.path()), advanced_regex)?;

        wait_for_text(&mut snapshot_rx, "Search text", 10).await?;

        send_chars("before", &event_sender);
        send_key(KeyCode::Tab, &event_sender);
        send_chars("after", &event_sender);
        send_key(KeyCode::Enter, &event_sender);

        wait_for_text(&mut snapshot_rx, "Still searching", 500).await?;

        wait_for_text(&mut snapshot_rx, "Search complete", 1000).await?;

        // Nothing should have changed yet
        assert_test_files!(
            &temp_dir,
            "dir1/file1.txt" => {
                "This is some test content before 123",
                "  with some spaces at the start",
                "and special ? characters 1! @@ # and number 890",
                "        some    tabs  and - more % special   **** characters ())",
            },
            "file2.py" => {
                "from datetime import datetime as dt, timedelta as td",
                "def mix_types(x=100, y=\"test\"): return f\"{x}_{y}\" if isinstance(x, int) else None",
                "class TestClass:",
                "    super_long_name_really_before_long_name_very_long_name = 123",
                "    return super_long_name_really_before_long_name_very_long_name",
                "test_dict = {\"key1\": [1,2,3], 123: \"num key\", (\"a\",\"b\"): True, \"before\": 1, \"test-key\": None}",
            },
        );

        send_key(KeyCode::Enter, &event_sender);

        wait_for_text(&mut snapshot_rx, "Success!", 1000).await?;

        // Verify that "before" has been replaced with "after"
        assert_test_files!(
            &temp_dir,
            "dir1/file1.txt" => {
                "This is some test content after 123",
                "  with some spaces at the start",
                "and special ? characters 1! @@ # and number 890",
                "        some    tabs  and - more % special   **** characters ())",
            },
            "file2.py" => {
                "from datetime import datetime as dt, timedelta as td",
                "def mix_types(x=100, y=\"test\"): return f\"{x}_{y}\" if isinstance(x, int) else None",
                "class TestClass:",
                "    super_long_name_really_after_long_name_very_long_name = 123",
                "    return super_long_name_really_after_long_name_very_long_name",
                "test_dict = {\"key1\": [1,2,3], 123: \"num key\", (\"a\",\"b\"): True, \"after\": 1, \"test-key\": None}",
            },
        );

        shutdown(event_sender, run_handle).await
    }
);

test_with_both_regex_modes!(
    test_search_and_replace_no_matches,
    |advanced_regex| async move {
        let temp_dir = &create_test_files!(
            "dir1/file1.txt" => {
                "This is some test content 123",
            },
        );

        let (run_handle, event_sender, mut snapshot_rx) =
            build_test_runner(Some(temp_dir.path()), advanced_regex)?;

        wait_for_text(&mut snapshot_rx, "Search text", 10).await?;

        send_chars("before", &event_sender);
        send_key(KeyCode::Tab, &event_sender);
        send_chars("after", &event_sender);
        send_key(KeyCode::Enter, &event_sender);

        wait_for_text(&mut snapshot_rx, "Still searching", 500).await?;

        wait_for_text(&mut snapshot_rx, "Search complete", 1000).await?;

        // Nothing should have changed yet
        assert_test_files!(
            &temp_dir,
            "dir1/file1.txt" => {
                "This is some test content 123",
            },
        );

        send_key(KeyCode::Enter, &event_sender);

        wait_for_text(&mut snapshot_rx, "Success!", 1000).await?;

        // Verify that nothing has changed
        assert_test_files!(
            &temp_dir,
            "dir1/file1.txt" => {
                "This is some test content 123",
            },
        );

        shutdown(event_sender, run_handle).await
    }
);

test_with_both_regex_modes_and_fixed_strings!(
    test_search_and_replace_empty_dir,
    |advanced_regex, fixed_strings| async move {
        let temp_dir = &create_test_files!();

        let (run_handle, event_sender, mut snapshot_rx) =
            build_test_runner(Some(temp_dir.path()), advanced_regex)?;

        wait_for_text(&mut snapshot_rx, "Search text", 10).await?;

        send_chars("before", &event_sender);
        send_key(KeyCode::Tab, &event_sender);
        send_chars("after", &event_sender);
        if fixed_strings {
            send_key(KeyCode::Tab, &event_sender);
            send_chars(" ", &event_sender); // Toggle on fixed strings
        }
        send_key(KeyCode::Enter, &event_sender);

        wait_for_text(&mut snapshot_rx, "Still searching", 500).await?;

        wait_for_text(&mut snapshot_rx, "Search complete", 1000).await?;

        assert_test_files!(&temp_dir);

        send_key(KeyCode::Enter, &event_sender);

        wait_for_text(&mut snapshot_rx, "Success!", 1000).await?;

        assert_test_files!(&temp_dir);

        shutdown(event_sender, run_handle).await
    }
);

test_with_both_regex_modes_and_fixed_strings!(
    test_search_and_replace_whole_words,
    |advanced_regex, fixed_strings| async move {
        let temp_dir = &create_test_files!(
            "dir1/file1.txt" => {
                "this is something",
                "some text someone abcsome123",
                "some",
                "dashes-some-text",
                "slashes and commas/some,text",
                "moresometext",
                "text some",
            },
            "file2.py" => {
                "print('Hello, some world!')",
            },
        );

        let (run_handle, event_sender, mut snapshot_rx) =
            build_test_runner(Some(temp_dir.path()), advanced_regex)?;

        wait_for_text(&mut snapshot_rx, "Search text", 10).await?;

        send_chars("some", &event_sender);
        send_key(KeyCode::Tab, &event_sender);
        send_chars("REPLACE", &event_sender);
        send_key(KeyCode::Tab, &event_sender);
        if fixed_strings {
            send_chars(" ", &event_sender); // Toggle on fixed strings
        }
        send_key(KeyCode::Tab, &event_sender);
        send_chars(" ", &event_sender); // Toggle on whole word matching
        send_key(KeyCode::Enter, &event_sender);

        wait_for_text(&mut snapshot_rx, "Still searching", 500).await?;

        wait_for_text(&mut snapshot_rx, "Search complete", 1000).await?;

        // Nothing should have changed yet
        assert_test_files!(
            &temp_dir,
            "dir1/file1.txt" => {
                "this is something",
                "some text someone abcsome123",
                "some",
                "dashes-some-text",
                "slashes and commas/some,text",
                "moresometext",
                "text some",
            },
            "file2.py" => {
                "print('Hello, some world!')",
            },
        );

        send_key(KeyCode::Enter, &event_sender);

        wait_for_text(&mut snapshot_rx, "Success!", 1000).await?;

        // Verify that "before" has been replaced with "after"
        assert_test_files!(
            &temp_dir,
            "dir1/file1.txt" => {
                "this is something",
                "REPLACE text someone abcsome123",
                "REPLACE",
                "dashes-REPLACE-text",
                "slashes and commas/REPLACE,text",
                "moresometext",
                "text REPLACE",
            },
            "file2.py" => {
                "print('Hello, REPLACE world!')",
            },
        );

        shutdown(event_sender, run_handle).await
    }
);

test_with_both_regex_modes!(
    test_search_and_replace_regex_capture_group,
    |advanced_regex| async move {
        let temp_dir = &create_test_files!(
            "phones.txt" => {
                "Phone: (020) 7123-4567",
                "Another: (0161) 4969-8523",
                "Different format: 020.7123.4567",
                "Also different: 020-7123-4567",
            },
        );

        let (run_handle, event_sender, mut snapshot_rx) =
            build_test_runner(Some(temp_dir.path()), advanced_regex)?;

        wait_for_text(&mut snapshot_rx, "Search text", 10).await?;

        send_chars(r"\((\d{3,4})\)\s(\d{4})-(\d{4})", &event_sender);
        send_key(KeyCode::Tab, &event_sender);
        send_chars("+44 $2 $1-$3", &event_sender);
        send_key(KeyCode::Enter, &event_sender);

        wait_for_text(&mut snapshot_rx, "Still searching", 500).await?;
        wait_for_text(&mut snapshot_rx, "Search complete", 1000).await?;

        // Nothing should have changed yet
        assert_test_files!(
            &temp_dir,
            "phones.txt" => {
                "Phone: (020) 7123-4567",
                "Another: (0161) 4969-8523",
                "Different format: 020.7123.4567",
                "Also different: 020-7123-4567",
            },
        );

        send_key(KeyCode::Enter, &event_sender);

        wait_for_text(&mut snapshot_rx, "Success!", 1000).await?;

        // Verify only matching phone numbers are reformatted
        assert_test_files!(
            &temp_dir,
            "phones.txt" => {
                "Phone: +44 7123 020-4567",
                "Another: +44 4969 0161-8523",
                "Different format: 020.7123.4567",
                "Also different: 020-7123-4567",
            },
        );

        shutdown(event_sender, run_handle).await
    }
);

#[tokio::test]
async fn test_search_and_replace_advanced_regex_negative_lookahead() -> anyhow::Result<()> {
    let temp_dir = &create_test_files!(
        "src/lib.rs" => {
            "fn process(mut data: Vec<u32>) {",
            "    let mut count = 0;",
            "    let total = 0;",
            "    let values = Vec::new();",
            "    let mut items = data.clone();",
            "    let result = compute(data);",
            "}",
            "",
            "fn compute(input: Vec<u32>) -> u32 {",
            "    let mut sum = 0;",
            "    let multiplier = 2;",
            "    let base = 10;",
            "    sum",
            "}",
        },
    );

    let (run_handle, event_sender, mut snapshot_rx) =
        build_test_runner(Some(temp_dir.path()), true)?;

    wait_for_text(&mut snapshot_rx, "Search text", 10).await?;

    // Match 'let' declarations that aren't mutable
    // Use negative lookbehind for function parameters and negative lookahead for mut
    send_chars(r"(?<!mut\s)let\s(?!mut\s)(\w+)", &event_sender);
    send_key(KeyCode::Tab, &event_sender);
    send_chars("let /* immutable */ $1", &event_sender);
    send_key(KeyCode::Enter, &event_sender);

    wait_for_text(&mut snapshot_rx, "Still searching", 500).await?;
    wait_for_text(&mut snapshot_rx, "Search complete", 1000).await?;

    // Nothing should have changed yet
    assert_test_files!(
        &temp_dir,
        "src/lib.rs" => {
            "fn process(mut data: Vec<u32>) {",
            "    let mut count = 0;",
            "    let total = 0;",
            "    let values = Vec::new();",
            "    let mut items = data.clone();",
            "    let result = compute(data);",
            "}",
            "",
            "fn compute(input: Vec<u32>) -> u32 {",
            "    let mut sum = 0;",
            "    let multiplier = 2;",
            "    let base = 10;",
            "    sum",
            "}",
        },
    );

    send_key(KeyCode::Enter, &event_sender);

    wait_for_text(&mut snapshot_rx, "Success!", 1000).await?;

    // Verify only non-mutable declarations are modified
    assert_test_files!(
        &temp_dir,
        "src/lib.rs" => {
            "fn process(mut data: Vec<u32>) {",
            "    let mut count = 0;",
            "    let /* immutable */ total = 0;",
            "    let /* immutable */ values = Vec::new();",
            "    let mut items = data.clone();",
            "    let /* immutable */ result = compute(data);",
            "}",
            "",
            "fn compute(input: Vec<u32>) -> u32 {",
            "    let mut sum = 0;",
            "    let /* immutable */ multiplier = 2;",
            "    let /* immutable */ base = 10;",
            "    sum",
            "}",
        },
    );

    shutdown(event_sender, run_handle).await
}
// TODO: add tests for using fixed strings
