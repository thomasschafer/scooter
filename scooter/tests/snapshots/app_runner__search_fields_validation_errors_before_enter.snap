---
source: scooter/tests/app_runner.rs
expression: snapshot
---
                                               scooter                                              
    ┌Search text (Error: Couldn't parse regex)───────────────────────────────────────────────┐     
    │(                                                                                       │     
    └────────────────────────────────────────────────────────────────────────────────────────┘     
    ┌Replace text────────────────────────────────────────────────────────────────────────────┐     
    │                                                                                        │     
    └────────────────────────────────────────────────────────────────────────────────────────┘     
    ┌───┐                                                                                          
    │   │ Fixed strings                                                                            
    └───┘                                                                                          
    ┌───┐                                                                                          
    │   │ Match whole word                                                                         
    └───┘                                                                                          
    ┌───┐                                                                                          
    │ X │ Match case                                                                               
    └───┘                                                                                          
    ┌Files to include (Error: Couldn't parse glob pattern)───────────────────────────────────┐     
    │*, {                                                                                    │     
    └────────────────────────────────────────────────────────────────────────────────────────┘     
    ┌Files to exclude (Error: Couldn't parse glob pattern)───────────────────────────────────┐     
    │{{                                                                                      │     
    └────────────────────────────────────────────────────────────────────────────────────────┘     
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
         <ret> jump to results / <tab> focus next / <C-r> reset / <C-h> help / <C-c> quit
